File: Go/Day04.go
package main

import AH "./adventhelper"

import (
	"strconv"
	"strings"
)

func check(s string) (bool, bool) {
	bounds := strings.Split(s, ",")
	lhs, rhs := strings.Split(bounds[0], "-"), strings.Split(bounds[1], "-")
	l1, _ := strconv.Atoi(lhs[0])
	l2, _ := strconv.Atoi(lhs[1])
	r1, _ := strconv.Atoi(rhs[0])
	r2, _ := strconv.Atoi(rhs[1])

	part1 := ((l1 <= r1) && (r2 <= l2)) || ((r1 <= l1) && (l2 <= r2))
	part2 := ((l2 >= r1) && (r2 >= l1))

	return part1, part2
}

func main() {
	space, _ := AH.ReadStrFile("../input/input04.txt")

	part1, part2 := 0, 0
	for _, s := range space {
		p1, p2 := check(s)
		if (p1) {
			part1 += 1
		}
		if (p2) {
			part2 += 1
		}
	}

	AH.PrintSoln(4, part1, part2)

	return
}

File: Rust/src/day04.rs
use crate::advent_helper::advent_helper::read_strs;
use crate::advent_helper::advent_helper::print_soln;

fn to_range(s: &str) -> (i64, i64)
{
	let v: Vec<i64> = s.split("-").map(|x| x.parse::<i64>().unwrap()).collect();
	return (v[0], v[1]);
}

fn check(s: &String) -> (bool, bool)
{
	let bounds: Vec<(i64, i64)> = s.split(",").map(|x| to_range(x)).collect();

	let p1 = ((bounds[0].0 <= bounds[1].0) && (bounds[1].1 <= bounds[0].1)) ||
	         ((bounds[1].0 <= bounds[0].0) && (bounds[0].1 <= bounds[1].1));
	let p2 = (bounds[0].1 >= bounds[1].0) && (bounds[1].1 >= bounds[0].0);

	return (p1, p2);
}

pub fn run()
{
	let space = read_strs("../input/input04.txt").unwrap();
	let part1: Vec<&String> = space.iter().filter(|x| check(x).0).collect();
	let part2: Vec<&String> = space.iter().filter(|x| check(x).1).collect();

	print_soln(4, part1.len(), part2.len());
	return;
}
